sklearn/utils/validation.py

```diff
@@ -1089,3 +1089,22 @@ def _allclose_dense_sparse(x, y, rtol=1e-7, atol=1e-9):
         return np.allclose(x, y, rtol=rtol, atol=atol)
     raise ValueError("Can only compare two sparse matrices, not a sparse "
                      "matrix and an array")
+
+
+def _validate_bad_defaults(obj):
+    if not hasattr(obj, "_bad_defaults"):
+        return
+
+    obj_values = {param: getattr(obj, param) for param in obj._bad_defaults}
+    bad_params = [param for param, value in obj_values.items()
+                  if value == 'warn']
+    if bad_params:
+        msg = ("There is no good default value for the following "
+               "parameters in {}. Please consult the documentation "
+               "on how to set them for your data.\n\t".format(
+                   obj.__class__.__name__))
+        msg += '\n\t'.join(["'{}' - using default value: {}".format(
+            param, obj._bad_defaults[param]) for param in bad_params])
+        warnings.warn(msg, UserWarning)
+    for param in bad_params:
+        setattr(obj, param, obj._bad_defaults[param])
```

Comment:
Aren't you setting in `fit` a parameter passed to `__init__` here?

Reply:
that's a good point. It can be fixed, but it may not be a bad idea not to fix it. Basically, if the user clones the object after fit, they won't have the warning again (which may be a good thing, if we don't want it to be too crowded). But if they clone before the fit, which is done in grid search cv for instance, they'll see the warning for every fit, if it's not setting the parameter.

I'm probably forgetting some of the reasons why we don't change anything given to `__init__`.

Reply:
That has to be fixed in some way, else the estimator can't pass `check_estimator`.

Maybe one solution would be to call the validation in `__init__` instead (we would have to relax some of the checks but that should be OK)